cdap-common/src/main/java/co/cask/cdap/common/discovery/ResolvingDiscoverable.java

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright © 2015 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package co.cask.cdap.common.discovery;
+
+import org.apache.twill.discovery.Discoverable;
+
+import java.net.Inet6Address;
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.net.UnknownHostException;
+
+/**
+ * Discoverable that resolves 0.0.0.0 to a routable interface.
+ */
+public class ResolvingDiscoverable implements Discoverable {
+
+  private final Discoverable discoverable;
+
+  private ResolvingDiscoverable(Discoverable discoverable) {
+    this.discoverable = discoverable;
+  }
+
+  public static ResolvingDiscoverable of(Discoverable discoverable) {
+    return new ResolvingDiscoverable(discoverable);
+  }
+
+  @Override
+  public String getName() {
+    return discoverable.getName();
+  }
+
+  @Override
+  public InetSocketAddress getSocketAddress() {
+    return resolve(discoverable.getSocketAddress());
+  }
+
+  private InetSocketAddress resolve(InetSocketAddress bindAddress) {
+    try {
+      if (Inet6Address.getByName("0:0:0:0:0:0:0:0").equals(bindAddress.getAddress())) {
```

Comment:
Start your services with `-Djava.net.preferIPv4Stack=true` or you can disable IPv6 on the host and reboot so you won't get any IPv6 addresses. However, you should check for both, and resolve IPv6 before IPv4, since that's what all modern OS (Linux, OSX, Windows even) do when both are supported.